text/0000-kadalu-thin-arbiter-support.md

```diff
@@ -0,0 +1,84 @@
+
+The challenge with Replica3 is, the cost of adding one more full set of data is high for users (1/3 increase in hardware cost, and 1/3 reduction in write speed). Hence the recommended solution was 'Arbiter' volume type in gluster, which reduced the 'storage cost' involved in 3rd copy, as it doesn't host any data, but still user has to manage 3 set of machines, and even in arbiter case, the latency requirements are high.
+
+
+So, the thin-arbiter feature is very useful for users who have just 2 data centers with required latency, but not the 3rd. With this feature, all such users can actually have the 3rd tie-breaker node in cloud without latency limitation.
+
+Even for kadalu users, if you have a stretched k8s cluster, which spans across 2 data centers (or 2 availability zones in Cloud), this feature will be useful. This helps by reducing the resources required for high-availability, and also the speed of write (1/2 instead of 1/3).
+
+The design details of thin-arbiter was discussed [here](https://github.com/gluster/glusterfs/issues/352) and [here](https://docs.gluster.org/en/latest/Administrator%20Guide/Thin-Arbiter-Volumes/). In this proposal we will talk about how we can add thin-arbiter volume type to kadalu.
+
+
+## Design
+
+Kadalu works with predefined volume templates for gluster. We need add thin-arbiter template, and also define a specific storage `Type` for this template. The same can be used for storage-config 'type' too. Already `Replica1` and `Replica3` namespace taken. We can choose 'Replica2' as it is indeed a replica2 setup.
+
+Proposed config file looks something like below:
+
+```yaml
+# File: storage-config.yaml
+---
+apiVersion: kadalu-operator.storage/v1alpha1
+kind: KadaluStorage
+metadata:
+ # This will be used as name of PV Hosting Volume
+  name: storage-pool-name
+spec:
+  type: Replica2       # Just to highlight your data is just in 2 places.
+  storage:
+    - node: kube1      # node name as shown in `kubectl get nodes`
+      device: /dev/vdc # Device to provide storage to all PVs
+    - node: kube2      # node name as shown in `kubectl get nodes`
+      device: /dev/vdc # Device to provide storage to all PVs
+    - tie-breaker:
+        deployment: local # other option is 'external'.
+        node: tie-breaker.kadalu.io
+        path: /mnt/tie-breaker/
+```
+
+
+In the PV spec, one would need to use the option as below:
+
+```yaml
+# -*- mode: yaml -*-
+---
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: pv1
+spec:
+  storageClassName: kadalu.replica2
+  accessModes:
+    - ReadWriteOnce # You can also provide 'ReadWriteMany' here
+  resources:
+    requests:
+      storage: 1Gi
+```
+
+The main definition which needs to be properly validated would be of `tie-breaker`. If it has field which says `deployment: local`, then kadalu operator should deploy the thin-arbiter locally in the node. If it is `external` then, we can assume it is handled outside.
```

Comment:
If two KadaluStorage chooses local tiebreaker then Operator will start two tiebreaker pods, The operator can check and reuse the tiebreaker node if it exists already.

Reply:
True. While that is possible, is it possible for admin to say `local` for one config and `external` for another config? That way, it would be simple in code :-D